### PR TITLE
ENHANCEMENT: Cache canCreate in CMSMain

### PR DESCRIPTION
**Pull request by unclecheese**
Resolves https://github.com/silverstripe/silverstripe-cms/issues/1977

## Benchmarks

(-) CMSMain cache
(-) SiteTree cache
* 15.5s wall time
* 32.1MB memory
* 1.45s response time

(+) CMSMain cache
(-) SiteTree cache
* 15.3s wal
* 30.6MB
* 1.2s

(+) CMSMain cache
(+) SiteTree cache
* 7.5s
* 0.9s
* 31MB

### CODE REVIEW

**Comment by unclecheese** [comment]
Could use some unit tests once overall approach is approved.

**Comment by dhensby** [comment]
This needs to go against `4` branch due to new APIs

File: code/Controllers/CMSMain.php [hunk 1]
```diff
@@ -379,6 +388,22 @@ public function getTabIdentifier()
         return 'edit';
     }
 
+    /**
+     * @param CacheInterface $cache
+     */
+    public function setHintsCache(CacheInterface $cache)
```
Comment (tractorcow) [inline code/Controllers/CMSMain.php]: The setter needs a getter.

Also please `return $this;` to make it fluent.

File: code/Controllers/CMSMain.php [hunk 2]
```diff
@@ -160,6 +160,15 @@ class CMSMain extends LeftAndMain implements CurrentPageIdentifier, PermissionPr
         'SiteTreeAsUL' => 'HTMLFragment',
     );
 
+    private static $dependencies = [
```
Comment (tractorcow) [inline code/Controllers/CMSMain.php]: Can you make the case of the property match the public getter / setter.

```php
'HintsCache'
```

File: code/Controllers/CMSMain.php [hunk 3]
```diff
@@ -974,19 +999,25 @@ public function Breadcrumbs($unlinked = false)
...
 
-        // Generate basic cache key. Too complex to encompass all variations
-        $cache = Injector::inst()->get(CacheInterface::class . '.CMSMain_SiteTreeHints');
-        $cacheKey = md5(implode('_', array(Security::getCurrentUser()->ID, implode(',', $cacheCanCreate), implode(',', $classes))));
-        if ($this->getRequest()->getVar('flush')) {
-            $cache->clear();
-        }
-        $json = $cache->get($cacheKey);
+        // Cache the JSON string representing allowed children
+        $hintsCacheKey = md5(implode('_', [$memberID, implode(',', $cacheCanCreate), $classList]));
```
Comment (tractorcow) [inline code/Controllers/CMSMain.php]: We don't really need two cache keys here; Can we consolidate this together?

File: code/Controllers/CMSMain.php [hunk 4]
```diff
@@ -974,19 +999,25 @@ public function Breadcrumbs($unlinked = false)
...
-
-        $cacheCanCreate = array();
-        foreach ($classes as $class) {
-            $cacheCanCreate[$class] = singleton($class)->canCreate();
+        $memberID = Security::getCurrentUser()->ID;
+        $cache = $this->hintsCache;
+        $classList = implode(',', $classes);
+
+        // Cache the results of canCreate()
+        $permissionsCacheKey = md5(implode('_', [$memberID, $classList]));
```
Comment (tractorcow) [inline code/Controllers/CMSMain.php]: $classlist isn't a variable, it's a fact (fixed) for each project. I know it was there earlier but it's probably redundant, now that we have `flush` to clear (should be done when adding new classes).

File: code/Model/SiteTree.php [hunk 5]
```diff
@@ -2940,6 +2970,15 @@ public static function reset()
         }
     }
 
+    /**
+     * Clear the creatableChildren cache on flush
+     */
+    public static function flush()
+    {
+        Injector::inst()->get(CacheInterface::class . '.SiteTree_CreatableChildren')
```
Comment (tractorcow) [inline code/Model/SiteTree.php]: This won't respect `$dependencies` config; Could we use `static::singleton()->getCreatableChildrenCache()` instead?

File: code/Model/SiteTree.php [hunk 6]
```diff
@@ -904,6 +917,14 @@ public function getParent()
         return null;
     }
 
+    /**
+     * @param CacheInterface $cache
+     */
+    public function setCreatableChildrenCache(CacheInterface $cache)
```
Comment (tractorcow) [inline code/Model/SiteTree.php]: Need a getter for this setter.

File: code/Model/SiteTree.php [hunk 7]
```diff
@@ -2512,6 +2533,26 @@ public function allowedChildren()
...
+            $children = array();
+            $candidates = static::page_type_classes();
+            foreach ($candidates as $childClass) {
+                $child = singleton($childClass);
+                if ($child->canCreate(null, array('Parent' => $this))) {
+                    $children[$childClass] = $child->i18n_singular_name();
+                }
+            }
+            $cache->set($cacheKey, $children);
+        }
```
Comment (tractorcow) [inline code/Model/SiteTree.php]: Is this method meant to return something?

File: code/Controllers/CMSMain.php [hunk 8]
```diff
@@ -974,19 +999,25 @@ public function Breadcrumbs($unlinked = false)
     public function SiteTreeHints()
     {
         $classes = SiteTree::page_type_classes();
-
-        $cacheCanCreate = array();
-        foreach ($classes as $class) {
-            $cacheCanCreate[$class] = singleton($class)->canCreate();
+        $memberID = Security::getCurrentUser()->ID;
+        $cache = $this->hintsCache;
```
Comment (tractorcow) [inline code/Controllers/CMSMain.php]: Once we add a getter we should use that instead of direct properties.

File: code/Controllers/CMSMain.php [hunk 9]
```diff
@@ -974,19 +999,25 @@ public function Breadcrumbs($unlinked = false)
     public function SiteTreeHints()
     {
         $classes = SiteTree::page_type_classes();
-
-        $cacheCanCreate = array();
-        foreach ($classes as $class) {
-            $cacheCanCreate[$class] = singleton($class)->canCreate();
+        $memberID = Security::getCurrentUser()->ID;
```
Comment (tractorcow) [inline code/Controllers/CMSMain.php]: Assumes getCurrentUser() isn't null.

File: code/Model/SiteTree.php [hunk 10]
```diff
@@ -2512,6 +2533,26 @@ public function allowedChildren()
         return $allowedChildren;
     }
 
+    public function creatableChildren()
```
Comment (tractorcow) [inline code/Model/SiteTree.php]: PHPDoc on return type and purpose please. :)

Could this be protected?

**Comment by tractorcow** [comment]
Good job overall. :)

Are you able to get some kind of performance metric on SiteTreeHints() and getTreeTitle() to see how much faster it is?

**Comment by unclecheese** [comment]
Thanks for those. Good points.

One glaring omission is that it's missing a cache-busting mechanism on Member/Group write. Here's the options, as I see it:

* We could make `CMSMain` and `SiteTree` implementors of `CacheFlusher` (`flushCache($ids = [])` ), but that would put the `flushCache($ids)` method in conflict with the native `DataObject::flushCache()` method. Possible solution: change `CacheFlusher` to define `flushBackendCache($ids)` or something along those lines.

* We create a new layer of abstraction and hand off the entire implementation to another service, which would scale better:

```
class CreateableChildrenCache implements CacheFlusher
{
  __construct(CacheInterface $cache)
  
  set($memberID, $siteTreeClass, $pageID)

  flushCache($memberIDs = []))  
}
```

**Comment by tractorcow** [comment]
Pretty damn awesome. :)

I had to re-run `phpcbf` over it again to fix linting issues, and added a few phpdoc fixes, plus some minor code cleanup. :)

It's all in one commit if you want to review @unclecheese.

Otherwise, can be merged on green.

File: _config/permissions.yml [hunk 11]
```diff
@@ -6,7 +6,6 @@ SilverStripe\Core\Injector\Injector:
     class: SilverStripe\Security\InheritedPermissions
     constructor:
       BaseClass: SilverStripe\CMS\Model\SiteTree
-      CacheService: '%$Psr\SimpleCache\CacheInterface.InheritedPermissions'
```
Comment (tractorcow) [inline _config/permissions.yml]: Why was this removed here?

**Comment by tractorcow** [comment]
I've squashed a bunch of work, and fixed some caching and service bugs.

File: code/Model/SiteTree.php [hunk 12]
```diff
@@ -1508,6 +1541,24 @@ public function flushCache($persistent = true)
...
 
+    /**
+     * Flushes the member specific cache for creatable children
+     * @param null $memberIDs
+     */
+    public function flushMemberCache($memberIDs = null)
+    {
+        $cache = static::singleton()->getCreatableChildrenCache();
+
+        if ($memberIDs) {
```
Comment (tractorcow) [inline code/Model/SiteTree.php]: I think you had this condition inverted
